dimod/views.py

```diff
@@ -72,3 +74,239 @@ class IndexValuesView(abc.ValuesView):
+
+    def copy(self):
+        if hasattr(self, '_data'):
+            # make a shallow copy of data.
+            return type(self)(self.value, self.data.copy())
+        else:
+            return type(self)(self.value)
+
+    @property
+    def data(self):
+        """dict: data associated with a variable or interaction."""
+        try:
+            return self._data
+        except AttributeError:
+            pass
+
+        self._data = data = {}
+        return data
+    
+
+class LinearView(abc.MutableMapping):
+    """Acts as a dictionary `{v: bias, ...}` for the linear biases.
+
+    The linear biases are stored in a dict-of-dicts format, where 'self loops'
+    store the linear biases.
+    So `{v: bias}` is stored `._adj = {v: {v: Bias(bias)}}`.
+
+    """
+
+    __slots__ = '_adj',
+
+    def __init__(self, bqm):
+        self._adj = bqm._adj
+
+    def __delitem__(self, v):
+        if v not in self:
+            raise KeyError
+        if len(self._adj[v]) > 1:
+            raise ValueError("there are interactions associated with {} that must be deleted first".format(v))
+        del self._adj[v]
+
+    def __getitem__(self, v):
+        return self._adj[v][v].value
+
+    def __iter__(self):
+        return iter(self._adj)
+
+    def __len__(self):
+        return len(self._adj)
+
+    def __setitem__(self, v, bias):
+        adj = self._adj
+        if v in adj:
+            adj[v][v].value = bias
+        else:
+            adj[v] = {v: Bias(bias)}
+
+    def __str__(self):
+        return str(dict(self))
+
+    def items(self):
+        return LinearItemsView(self)
+
+
+class LinearItemsView(abc.ItemsView):
+    """Faster items iteration for LinearView."""
+
+    __slots__ = ()
+
+    def __iter__(self):
+        for v, neighbours in self._mapping._adj.items():
+            yield v, neighbours[v].value
+
+
+class QuadraticView(abc.MutableMapping):
+    """Acts as a dictionary `{(u, v): bias, ...}` for the quadratic biases.
+
+    The quadratic biases are stored in a dict-of-dicts format. So `{(u, v): bias}` is stored as
+    `._adj = {u: {v: Bias(bias)}, v: {u: Bias(bias)}}`.
+
+    """
+
+    __slots__ = '_adj',
+
+    def __init__(self, bqm):
+        self._adj = bqm._adj
+
+    def __delitem__(self, interaction):
+        u, v = interaction
+        if u == v:
+            raise KeyError('{} is not an interaction'.format(interaction))
+        adj = self._adj
+        del adj[v][u]
+        del adj[u][v]
+
+    def __getitem__(self, interaction):
+        u, v = interaction
+        if u == v:
+            raise KeyError
+        return self._adj[u][v].value
+
+    def __iter__(self):
+        seen = set()
+        adj = self._adj
+        for u, neigh in adj.items():
+
+            for v in neigh:
+                if u == v:
+                    # not adjacent to itself
+                    continue
+
+                if v not in seen:
+                    yield (u, v)
+
+            seen.add(u)
```

Comment:
It would be nice if iteration would have a guaranteed/stable order, i.e. if it would step through sorted pairs of a triangular matrix. That would also resolve the O(N^3) time complexity this iterator has.